pkg/providers/v1/aws_test.go

```diff
@@ -3953,3 +4023,1093 @@ func TestIsAWSErrorInstanceNotFound(t *testing.T) {
+				assert.Error(t, err)
+				return
+			}
+			assert.NoError(t, err)
+
+			// Verify that the rules include the ICMP permission for MTU discovery
+			foundMTURule := false
+			for _, rule := range tc.rules {
+				if aws.ToString(rule.IpProtocol) == "icmp" &&
+					aws.ToInt32(rule.FromPort) == 3 &&
+					aws.ToInt32(rule.ToPort) == 4 {
+					foundMTURule = true
+					break
+				}
+			}
+			assert.True(t, foundMTURule, "MTU discovery rule should be added")
+
+			// Verify the ec2SourceRanges were properly set
+			for _, rule := range tc.rules {
+				if aws.ToString(rule.IpProtocol) == "icmp" {
+					assert.Equal(t, tc.ec2SourceRanges, rule.IpRanges)
+				}
+			}
+		})
+	}
+}
+
+func TestCreateSecurityGroup(t *testing.T) {
+	awsServices := newMockedFakeAWSServices(TestClusterID)
+	cfg := config.CloudConfig{}
+	cfg.Global.NLBSecurityGroupMode = config.NLBSecurityGroupModeManaged
+	c, _ := newAWSCloud(cfg, awsServices)
+
+	testCases := []struct {
+		name           string
+		sgName         string
+		sgDescription  string
+		additionalTags map[string]string
+		expectGroupID  string
+		expectError    bool
+	}{
+		{
+			name:          "successful security group creation",
+			sgName:        "test-sg",
+			sgDescription: "test security group",
+			additionalTags: map[string]string{
+				"key1": "value1",
+				"key2": "value2",
+			},
+			expectGroupID: "sg-123456",
+			expectError:   false,
+		},
+		{
+			name:           "empty security group name",
+			sgName:         "",
+			sgDescription:  "test security group",
+			additionalTags: map[string]string{},
+			expectGroupID:  "",
+			expectError:    true,
+		},
+		{
+			name:           "nil additional tags",
+			sgName:         "test-sg-2",
+			sgDescription:  "test security group 2",
+			additionalTags: nil,
+			expectGroupID:  "sg-123456",
+			expectError:    false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Mock DescribeSecurityGroups for ensureSecurityGroup (no existing SG)
+			awsServices.ec2.(*MockedFakeEC2).On("DescribeSecurityGroups", &ec2.DescribeSecurityGroupsInput{
+				Filters: []ec2types.Filter{
+					{
+						Name:   aws.String("group-name"),
+						Values: []string{tc.sgName},
+					},
+					{
+						Name:   aws.String("vpc-id"),
+						Values: []string{""},
+					},
+				},
+			}).Return([]ec2types.SecurityGroup{}, nil).Maybe()
+
+			// Mock CreateSecurityGroup with a simpler matcher
+			awsServices.ec2.(*MockedFakeEC2).On("CreateSecurityGroup", mock.AnythingOfType("*ec2.CreateSecurityGroupInput")).Return(&ec2.CreateSecurityGroupOutput{
+				GroupId: aws.String(tc.expectGroupID),
+			}, nil).Maybe()
+
+			groupID, err := c.createSecurityGroup(context.TODO(), tc.sgName, tc.sgDescription, tc.additionalTags)
+
+			if tc.expectError {
+				assert.Error(t, err)
+				assert.Empty(t, groupID)
+				return
+			}
+			assert.NoError(t, err)
+			assert.Equal(t, tc.expectGroupID, groupID)
+		})
+	}
+}
+
+func TestEnsureNLBSecurityGroup(t *testing.T) {
```

Comment:
FWIW/reviewer: I the next round I will join the following test functions to `TestEnsureNLBSecurityGroup` that was originally isolated, so it could be better for maintenance, however the complexity of setup would increase a bit due mocking instances:
- TestEnsureNLBSecurityGroupEdgeCases
- TestEnsureNLBSecurityGroupBYOSGValidation